kernel/thread/worker.c

```diff
@@ -89,13 +90,20 @@ void thd_worker_wakeup(kthread_worker_t *worker) {
 }
 
 void thd_worker_destroy(kthread_worker_t *worker) {
+    uint32_t flags;
+    kthread_t *thd;
+
     assert(worker != NULL);
 
+    flags = irq_disable();
+
     worker->quit = true;
     genwait_wake_one(worker);
+    thd = worker->thd;
 
-    thd_join(worker->thd, NULL);
-    free(worker);
```

Comment:
When a worker is destroyed, `thd_worker_destroy` attempts to wake the worker then `thd_join` it. In the case of a detached thread or trying to join oneself (worker's routine calls `thd_shutdown`), the `thd_join` will fail and we may then call the `free(worker)` before the worker's routine has started. This would lead to using worker after free.